### PR TITLE
Argument validation in mmap and munmap functions.

### PR DESCRIPTION
**Pull request by abujalski**
This change adds mmamp and munmap functions argument validations to fix tests from http://posixtest.sourceforge.net/
    
Following validations were added:
- validation of munmap arguments. munmap/8-1 and munmap/9-1 tests fixed,
- Checking if we have write access when prot is set to PROT_WRITE in mmap. mmap/6-4 test fixed,
- checking if addr is a multiple of the page size when flag MAP_FIXED is set. mmap/9-1 test fixed,
- Changed condition (fd == -1) to checking if MAP_ANONYMOUS flag is set in mmap2 syscall. mmap/19-1 test fixed.

### CODE REVIEW

File: src/library_fs.js [hunk 1]
```diff
@@ -1204,7 +1204,11 @@ mergeInto(LibraryManager.library, {
       stream.stream_ops.allocate(stream, offset, length);
     },
     mmap: function(stream, buffer, offset, length, position, prot, flags) {
-      // TODO if PROT is PROT_WRITE, make sure we have write access
+      if ((prot & {{{ cDefine('PROT_WRITE') }}}) !== 0
+          && (flags & {{{ cDefine('MAP_PRIVATE')}}}) === 0
```
Comment (kripken) [inline src/library_fs.js]: why is private relevant here?

File: src/library_syscall.js [hunk 2]
```diff
@@ -446,6 +446,7 @@ var SyscallsLibrary = {
   },
   __syscall91: function(which, varargs) { // munmap
     var addr = SYSCALLS.get(), len = SYSCALLS.get();
+    if (addr < 0 || len == 0) return -ERRNO_CODES.EINVAL;
```
Comment (kripken) [inline src/library_syscall.js]: why would `addr` be less than zero? (is this a check for an address greater than 2GB?)

File: src/library_syscall.js [hunk 3]
```diff
@@ -973,7 +974,12 @@ var SyscallsLibrary = {
     off <<= 12; // undo pgoffset
     var ptr;
     var allocated = false;
-    if (fd === -1) {
+
+    if ((flags & {{{ cDefine('MAP_FIXED') }}}) !== 0 && (addr % PAGE_SIZE) !== 0) {
```
Comment (kripken) [inline src/library_syscall.js]: I'm not sure what this is checking? A comment could be helpful - perhaps in the other places too.

File: src/library_fs.js (hunk 1 above)
Comment (abujalski) [reply src/library_fs.js]: According to POSIX spec yes:

"If PROT_WRITE is specified, the application shall ensure that it has opened the file descriptor fildes with write permission unless MAP_PRIVATE is specified in the flags parameter as described below."
http://pubs.opengroup.org/onlinepubs/9699919799/

I'll comment that.

File: src/library_syscall.js (hunk 2 above)
Comment (abujalski) [reply src/library_syscall.js]: This check was based on munmap/8-1 test from POSIX test suite which passes -1 as `addr` argument and expects EINVAL.

I'll reconsider this check.

**Comment by abujalski** [comment]
Oops, these *.orig files were unintended. Removed them in https://github.com/emscripten-core/emscripten/pull/8633/commits/43fcce7707af84dfe995df60a4fd80e8456e6fad

Added tests for mmap in https://github.com/emscripten-core/emscripten/pull/8633/commits/ac1fd225b8e32f77be4393dd0aac8d458940f8c5 & https://github.com/emscripten-core/emscripten/pull/8633/commits/c514fbedff0c62a8b58fb1a4969eec39c1b7378a

File: src/library_syscall.js (hunk 3 above)
Comment (abujalski) [reply src/library_syscall.js]: Added comment in https://github.com/emscripten-core/emscripten/pull/8633/commits/43fcce7707af84dfe995df60a4fd80e8456e6fad

File: src/library_syscall.js (hunk 2 above)
Comment (abujalski) [reply src/library_syscall.js]: Changed to check addr != MAP_FAILED in https://github.com/emscripten-core/emscripten/pull/8633/commits/43fcce7707af84dfe995df60a4fd80e8456e6fad

File: src/library_fs.js (hunk 1 above)
Comment (abujalski) [reply src/library_fs.js]: Added comment in https://github.com/emscripten-core/emscripten/pull/8633/commits/43fcce7707af84dfe995df60a4fd80e8456e6fad

**Comment by kripken** [comment]
Looks good, thanks! Just a question about testing - does this need to be in the `browser` test suite? If not, then it's much faster in the `other` suite. You would need to use `--embed-file` install of preloading, since preloading only works in the browser, I believe, but are there other issues?

**Comment by abujalski** [comment]
Actually, I haven't thought of adding this to other test suite than `browser`, as I saw `test_mmap_file` there. I'll try with `other` test suite.

File: tests/test_other.py [hunk 4]
```diff
@@ -9198,3 +9198,13 @@ def test(code):
...
+
+  def test_mmap_and_munmap(self):
+    cmd = [PYTHON, EMCC, path_from_root('tests', 'mmap_and_munmap.c')]
+    for f in ['data_ro.dat', 'data_rw.dat']:
+        create_test_file(f, 'Test file')
+        cmd.extend(['--embed-file', f])
+    run_process(cmd)
+    result = run_js('a.out.js')
+    self.assertContained('failures: 0', result)
+    self.assertNotContained('FAILED', result)
```
Comment (kripken) [inline tests/test_other.py]: in `other`, it's easier to use `self.do_other_test`.

That can only check for existing text, but I think it's enough to look for `failures: 0`?

Comment (abujalski) [reply tests/test_other.py]: Done.

**Comment by kripken** [comment]
Looks good, thanks @abujalski!

**Pull request merged**